design/ratelimit-design-new.md

```diff
@@ -0,0 +1,180 @@
+      port: 8081
+  timeoutPolicy:
+    response: 50ms
+```
+
+Second, a new block will be added to the Contour config file to identify the RLS extension service and provide some additional parameters:
+```yaml
+rateLimitService:
+  # extensionService contains the namespace/name of the RLS extension service.
+  extensionService: projectcontour/ratelimit
+  # domain is passed to the RLS for all rate limit requests.
+  domain: contour
+  # denyOnFailure defines whether to fail open or closed. If true, then if the RLS
+  # cannot be reached or does not return a valid rate limit decision within the
+  # specified timeout, the client will receive a 429 response for their request.
+  denyOnFailure: true
+```
+
+### Rate Limit Flows
+
+- First, Envoy applies local rate limiting to incoming requests.
+- If there are no tokens left in the relevant token bucket(s), a `429 Too Many Requests` response is returned.
+- If there was no local rate limiting defined, or there is a token available, Envoy proceeds to global rate limiting.
+- 1+ descriptors, each of which is an ordered list of key/value pairs, are generated for the request based on the global rate limit configuration.
+- The descriptors are sent to the external RLS via gRPC.
+- The external RLS makes a rate limit decision for the request, based on the descriptors it receives.
+- The external RLS returns a response indicating whether the client request should be rate-limited or not.
+- If the request should be rate limited, a `429 Too Many Requests` response is returned to the client along with the `x-envoy-ratelimited` header.
+- If the request should not be rate limited, it is routed to the appropriate upstream cluster and proceeds as normal.
+
+## Alternatives Considered
+
+### Tight integration with Lyft RLS
+Lyft has open-sourced a [reference external RLS implementation](https://github.com/envoyproxy/ratelimit/).
+We considered tightly integrating with this specific RLS implementation, to provide a more streamlined UX for configuring rate limits.
+A CRD could be used -- either `HTTPProxy` or a stand-alone one -- to define actual rate limits, and a controller could convert the CRDs into configuration in the Lyft RLS format.
+This option was discarded (for now) because we have potential users who are interested in using other rate limiting services, so just providing an integration with the Lyft implementation is not sufficient. 
+Future work could still be done to enable rate limits to be defined via CRD (as part of `HTTPProxy`, or stand-alone) and automatically configured with 1+ underlying RLS implementations.
+
+### Contour as external RLS
+Contour could function as an external RLS itself.
+This would put Contour in the data path for requests.
+It would simplify deployment and configuration for the user, at a cost of significant additional complexity for Contour.
+
+## Compatibility
+Rate limiting (both local and global) will be an optional, opt-in feature for Contour users.
```

Comment:
If you looked at how other proxies or CRDs express rate limits, maybe make some comparison notes here.